### PR TITLE
os/pm: Extend pm driver for pm suspend count

### PR DESCRIPTION
**Pull request by gSahitya-samsung**
This commit extends pm driver for pm suspend count. It adds API to get suspend
count of given domain id inside pm driver.

### CODE REVIEW

**Review by samsung-singh — APPROVED** [review]
added pm driver for pm suspend count and an API to get suspend count by domain id, good to merge.

File: os/drivers/pm/pm.c [hunk 1]
```diff
@@ -73,6 +73,7 @@ static ssize_t pm_write(FAR struct file *filep, FAR const char *buffer, size_t l
  *   PMIOC_SLEEP - to make board sleep for given time
  *   PMIOC_TIMEDSUSPEND - to suspend a pm state for given time duration
  *   PMIOC_TUNEFREQ - for changing the operating frequency of the core to save power
+ *   PMIOC_SUSPEND_COUNT - to get suspend count of pm domain
```
Comment (ritesh55555) [inline os/drivers/pm/pm.c:76]: Can you add command description for PMIOC_DOMAIN_REGISTER and PMIOC_METRICS

**Comment by sunghan-chang** [comment]
Are you able to leave usage of this new API? Why do you add this?

**Comment by gSahitya-samsung** [comment]
> Are you able to leave usage of this new API? Why do you add this?

There was request by @ewoodev to add this API. The application need suspend count information for debugging purpose.

**Review by r-prabu — DISMISSED** [review]
Getting the pm suspend count implementation is fine.

File: os/pm/pm_suspendcount.c [hunk 2]
```diff
@@ -77,25 +77,23 @@
...
- *
- * Assumptions:
- *   This function may be called from an interrupt handler.
+ *   non-negative interger: the suspend count of domain
+ *   ERROR: for invalid domain_id
  *
  ****************************************************************************/
 
-uint16_t pm_suspendcount(int domain_id)
+int pm_suspendcount(int domain_id)
```
Comment (ewoodev) [inline os/pm/pm_suspendcount.c:91]: I think, Return type should be same uint16_t

File: os/drivers/pm/pm.c [hunk 3]
```diff
@@ -128,6 +130,9 @@ static int pm_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
 			}
 		}
 		break;
+	case PMIOC_SUSPEND_COUNT:
+		ret = pm_suspendcount((int)arg);
```
Comment (ewoodev) [inline os/drivers/pm/pm.c:134]: How about using return for only error checking?

File: os/drivers/pm/pm.c (hunk 1 above)
Comment (gSahitya-samsung) [reply os/drivers/pm/pm.c:78]: Done

File: os/pm/pm_suspendcount.c (hunk 2 above)
Comment (gSahitya-samsung) [reply os/pm/pm_suspendcount.c:91]: We need `int` type to return `ERROR`

**Review by neel-samsung — APPROVED** [review]

**Review by ewoodev — APPROVED** [review]

**Review by sunghan-chang — DISMISSED** [review]

File: os/pm/pm_suspendcount.c [hunk 4]
```diff
@@ -77,25 +77,23 @@
...
 {
-	/* Get a convenience pointer to minimize all of the indexing */
-
-	DEBUGASSERT(domain_id >= 0 && domain_id < CONFIG_PM_NDOMAINS);
-
+	/* Check if domain_id is valid */
+	if (pm_check_domain(domain_id) != OK) {
+		return ERROR;
```
Comment (kishore-sn) [inline os/pm/pm_suspendcount.c:95]: How about returning EINVAL ?

**Comment by kishore-sn** [comment]
I think the commit description is wrong. You are not adding new pm driver. Just adding new ioctl for pm driver.

File: os/drivers/pm/pm.c (hunk 3 above)
Comment (gSahitya-samsung) [reply os/drivers/pm/pm.c:148]: `pm_check_domain` already setting `EINVAL` https://github.com/Samsung/TizenRT/blob/master/os/pm/pm_domain_register.c#L62 and we are returning it from `ioctl` call https://github.com/Samsung/TizenRT/blob/master/os/drivers/pm/pm.c#L147

File: os/pm/pm_suspendcount.c (hunk 4 above)
Comment (gSahitya-samsung) [reply os/pm/pm_suspendcount.c:95]: `pm_check_domain` already setting `EINVAL` https://github.com/Samsung/TizenRT/blob/master/os/pm/pm_domain_register.c#L62 and we are returning it from `ioctl` call https://github.com/Samsung/TizenRT/blob/master/os/drivers/pm/pm.c#L147

File: os/include/tinyara/pm/pm.h [hunk 5]
```diff
@@ -482,20 +482,18 @@ int pm_timedsuspend(int domain_id, unsigned int milliseconds);
...
  *
  * Input Parameters:
  *   domain_id - The domain ID of the PM activity
  *
  * Returned Value:
- *   Current pm stay count
- *
- * Assumptions:
- *   This function may be called from an interrupt handler.
+ *   non-negative interger: the suspend count of domain
```
Comment (kishore-sn) [inline os/include/tinyara/pm/pm.h:491]: Pls use capital letter for first character

File: os/pm/pm_suspendcount.c [hunk 6]
```diff
@@ -77,25 +77,23 @@
...
  *
  * Input Parameters:
  *   domain_id - The domain ID of the PM activity
  *
  * Returned Value:
- *   Current pm stay count
- *
- * Assumptions:
- *   This function may be called from an interrupt handler.
+ *   non-negative interger: the suspend count of domain
```
Comment (kishore-sn) [inline os/pm/pm_suspendcount.c:86]: Capital letter

File: os/include/tinyara/pm/pm.h (hunk 5 above)
Comment (gSahitya-samsung) [reply os/include/tinyara/pm/pm.h]: Done

File: os/pm/pm_suspendcount.c (hunk 6 above)
Comment (gSahitya-samsung) [reply os/pm/pm_suspendcount.c]: Done

**Review by kishore-sn — APPROVED** [review]

**Pull request merged**